pkg/controller/master/upgrade/upgrade_controller.go

```diff
@@ -161,6 +162,32 @@ func (h *upgradeHandler) OnChanged(key string, upgrade *harvesterv1.Upgrade) (*h
 			return h.upgradeClient.Update(toUpdate)
 		}
 
+		logrus.Info("Check minimum upgradable version")
+		minUpgradableVersion, err := repo.getMinUpgradableVersionStr()
+		if err != nil {
+			return nil, err
+		}
+		if minUpgradableVersion == "" {
+			logrus.Debug("No minimum upgradable version specified, continue the upgrading")
+		} else {
+			constraint := fmt.Sprintf(">= %s", minUpgradableVersion)
+
+			c, err := semverv3.NewConstraint(constraint)
+			if err != nil {
+				return nil, err
+			}
+			v, err := semverv3.NewVersion(upgrade.Status.PreviousVersion)
+			if err != nil {
+				return nil, err
+			}
+
+			if a := c.Check(v); !a {
+				reason := fmt.Sprintf("The current version %s is less than the minimum upgradable version %s.", upgrade.Status.PreviousVersion, minUpgradableVersion)
+				setUpgradeCompletedCondition(toUpdate, StateFailed, corev1.ConditionFalse, reason, "Current version not supported.")
```

Comment:
Here is what's in the condition when there is an error:

```
  conditions:
  - lastUpdateTime: "2022-09-26T10:32:07Z"
    message: Current version not supported.
    reason: The current version v1.1.0-dev1 is less than the minimum upgradable version
      v1.0.3.
    status: "False"
    type: Completed
```

The string "The current version v1.1.0-dev1 is less than the minimum upgradable version v1.0.3." is better to stay in the `message` field. One reason is that GUI displays the message first then the reason. Currently, it's not very informative:

<img width="315" alt="Screen Shot 2022-09-26 at 9 56 43 PM" src="https://user-images.githubusercontent.com/1691518/192296303-54149c81-f426-41d7-ba3e-36d6deab83a0.png">

We can check Nodes as an example; there are reasons like `KubeletHasSufficientMemory` and `KubeletHasNoDiskPressure`.

The existing setUpgradeCompletedCondition`calls put the error message as `reason` (my bad); I think we need to refactor them.

Reply:
Swapped the reason and message. Thanks!

Reply:
Now the error message is like:

<img width="579" alt="Screen Shot 2022-09-27 at 14 17 35" src="https://user-images.githubusercontent.com/1827717/192447687-e5ab19a2-3548-4eb6-970b-fa6e212736f7.png">

```bash
harv-2431:~ # k -n harvester-system get upgrades hvst-upgrade-rhnv6 -o jsonpath='{.status.conditions}' | jq .
[
  {
    "lastUpdateTime": "2022-09-27T06:16:15Z",
    "message": "The current version v1.1.0 is less than the minimum upgradable version v1.1.1.",
    "reason": "Current version not supported.",
    "status": "False",
    "type": "Completed"
  },
  {
    "lastUpdateTime": "2022-09-27T06:15:07Z",
    "status": "True",
    "type": "ImageReady"
  },
  {
    "lastUpdateTime": "2022-09-27T06:16:14Z",
    "status": "True",
    "type": "RepoReady"
  }
]
```